File: Data Structures & Algorithms/max-water-container/submission-2.py
class Solution:
    def maxArea(self, h: List[int]) -> int:
        # res = 0
        # for i in range(len(h)):
        #     for j in range(i+1, len(h)):
        #         res = max(res, min(h[i], h[j]) * (j-i) )
        # return res
        
        res = 0
        l,r = 0,len(h)-1
        while l < r:
            res = max( res  , min(h[l] , h[r]) * (r-l) )
            if h[l] <= h[r]:
                l = l+1
            else:
                r = r-1
        return res

File: Data Structures & Algorithms/binary-tree-diameter/submission-2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    private int height(TreeNode root)
    {
        if(root == null)
            return 0;
        return 1 + Math.max(height(root.left), height(root.right));
    }
    private int check(TreeNode root, int res)
    {
        if(root == null)
            return 0;
        int cur = height(root.left)+height(root.right);
        res = Math.max(cur, res);
        res = Math.max(  Math.max(check(root.left, res) , check(root.right,res)) , res );
        return res;
    }
    public int diameterOfBinaryTree(TreeNode root) {
        if(root == null)
            return 0;
        return check(root,0);
    }
}
